### PR TITLE
#137156981 Digital Products

### PR DESCRIPTION
**Pull request by fuzzybrows**
#### What does this PR do?
Enables the ability to add and subsequently purchase digital content

 #### Description of Task to be completed?
Store owners can now upload and buyers purchase digital products of the Book, Software, Music, and Video categories

 #### How should this be manually tested?
* Create a New product and Select the Digital Products toggle.
* Go on to upload the Digital Product and Publish
* Sign in as a customer and select the product. 
* Add Product to cart and checkout.
* A download button will be shown to download the purchased content.

 #### What are the relevant pivotal tracker stories?
#137156981 Digital Products

### CODE REVIEW

File: imports/plugins/included/product-detail-simple/client/components/variantList.js [hunk 1]
```diff
@@ -148,7 +166,9 @@ VariantList.propTypes = {
   childVariants: PropTypes.arrayOf(PropTypes.object),
   displayPrice: PropTypes.func,
   editable: PropTypes.bool,
+  isDigital: PropTypes.any,
```
Comment (adebayo-maborukoje) [inline imports/plugins/included/product-detail-simple/client/components/variantList.js:169]: is this not suppose to be a boolean?

File: server/methods/core/cart.js [hunk 2]
```diff
@@ -343,6 +344,9 @@ Meteor.methods({
         "_id": cart._id,
         "items.variants._id": variantId
       }, {
+        $addToSet: {
+          isDigital: isDigital
```
Comment (adebayo-maborukoje) [inline server/methods/core/cart.js:348]: whats the  difference between this n line  382

File: imports/plugins/included/product-detail-simple/client/components/variantList.js (hunk 1 above)
Comment (fuzzybrows) [reply imports/plugins/included/product-detail-simple/client/components/variantList.js:169]: Yes However I had to initialize the state as an empty string and Its also one of the conditions on line 131

File: server/methods/core/cart.js (hunk 2 above)
Comment (fuzzybrows) [reply server/methods/core/cart.js:348]: This is for the different cart instances. first it adds the isDigital property if the variant already exists in cart, and the second time if it's a new instance.

File: imports/plugins/core/orders/client/templates/list/ordersList.js [hunk 3]
```diff
@@ -10,6 +11,45 @@ Template.dashboardOrdersList.helpers({
   orderStatus() {
     return (this.workflow.status === "coreOrderCompleted");
   },
+  showDigitalFileDownload() {
+    const productId = this.items[0].productId;
+    const sub = Meteor.subscribe("Product", productId);
+    if (sub.ready()) {
+      const product = Products.findOne(productId);
+
+      if (product.digitalInfo.category === "audio") {
```
Comment (adebayo-maborukoje) [inline imports/plugins/core/orders/client/templates/list/ordersList.js]: These methods are repetitive. keep it DRY.

File: imports/plugins/included/product-detail-simple/client/components/digitalProducts.css [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+input:checked + .slider {
+  background-color: #2196F3;
+}
+
+input:focus + .slider {
+  box-shadow: 0 0 1px #2196F3;
+}
+
+.switch-middle {
+  text-align:center;
```
Comment (adebayo-maborukoje) [inline imports/plugins/included/product-detail-simple/client/components/digitalProducts.css]: let there be space between the colon and `center`.
the same thing applies to the next line.

File: imports/plugins/included/product-detail-simple/client/components/productDetail.js [hunk 5]
```diff
@@ -81,6 +102,176 @@ class ProductDetail extends Component {
...
+          const digitalInfo = {
+            fileId: file._id,
+            category: this.state.categoryValue
+          };
+
+          this.props.onProductFieldChange(this.props.product._id, "productFileId", file._id);
+          this.props.onProductFieldChange(this.props.product._id, "digitalInfo", digitalInfo);
+          return Alerts.toast("File successfully uploaded", "success");
+        }
+        return null;
```
Comment (adebayo-maborukoje) [inline imports/plugins/included/product-detail-simple/client/components/productDetail.js]: why are you returning `null` here?

File: imports/plugins/included/product-detail-simple/client/components/productDetail.js [hunk 6]
```diff
@@ -81,6 +102,176 @@ class ProductDetail extends Component {
...
+      });
+      return true;
+    }
+
+    if (this.state.categoryValue === "video") {
```
Comment (adebayo-maborukoje) [inline imports/plugins/included/product-detail-simple/client/components/productDetail.js]: All the methods below are repetitive as well. we can create one method and re use it.

File: imports/plugins/included/product-detail-simple/client/components/productDetail.js [hunk 7]
```diff
@@ -81,6 +102,176 @@ class ProductDetail extends Component {
...
+            category: this.state.categoryValue
+          };
+
+          this.props.onProductFieldChange(this.props.product._id, "productFileId", file._id);
+          this.props.onProductFieldChange(this.props.product._id, "digitalInfo", digitalInfo);
+          return Alerts.toast("File successfully uploaded", "success");
+        }
+        return null;
+      });
+      return true;
```
Comment (adebayo-maborukoje) [inline imports/plugins/included/product-detail-simple/client/components/productDetail.js]: why are you returning `true` on this line?

File: imports/plugins/included/product-detail-simple/client/components/productDetail.js [hunk 8]
```diff
@@ -170,6 +363,7 @@ class ProductDetail extends Component {
                   !this.props.editable ?
                   <AddToCartButton
                     cartQuantity={this.props.cartQuantity}
+                    inputType={this.state.cartQuantity}
```
Comment (adebayo-maborukoje) [inline imports/plugins/included/product-detail-simple/client/components/productDetail.js]: This should be renamed to reflect the true function of the state.

File: lib/collections/collectionFS.js [hunk 9]
```diff
@@ -64,3 +64,113 @@ export const Media = new FS.Collection("Media", {
...
+export const Software = new FS.Collection("Software", {
+  stores: [new FS.Store.GridFS("software", {
+    transformWrite: function (fileObj, readStream, writeStream) {
+      readStream.pipe(writeStream);
+    }
+  })
+  ]
+});
+
+Audio.allow({
```
Comment (adebayo-maborukoje) [inline lib/collections/collectionFS.js]: refactor this code 
1. Keep it DRY
2. ensure the function is actually doing something if at all it is needed.

File: server/publications/collections/media.js [hunk 10]
```diff
@@ -30,3 +30,71 @@ Meteor.publish("Media", function (shops) {
     }
   });
 });
+
+Meteor.publish("audio", function (productId) {
```
Comment (adebayo-maborukoje) [inline server/publications/collections/media.js]: This should also follow the DRY pattern

File: imports/plugins/core/orders/client/templates/list/ordersList.js (hunk 3 above)
Comment (fuzzybrows) [reply imports/plugins/core/orders/client/templates/list/ordersList.js]: Done, abstracted using a function

File: imports/plugins/included/product-detail-simple/client/components/digitalProducts.css (hunk 4 above)
Comment (fuzzybrows) [reply imports/plugins/included/product-detail-simple/client/components/digitalProducts.css]: Fixed.

File: imports/plugins/included/product-detail-simple/client/components/productDetail.js (hunk 5 above)
Comment (fuzzybrows) [reply imports/plugins/included/product-detail-simple/client/components/productDetail.js]: I'ts basically to satisfy the practice that a function must always return something.
I've changed this to return a message instead.

File: imports/plugins/included/product-detail-simple/client/components/productDetail.js (hunk 6 above)
Comment (fuzzybrows) [reply imports/plugins/included/product-detail-simple/client/components/productDetail.js]: Done, abstracted using a function

File: imports/plugins/included/product-detail-simple/client/components/productDetail.js (hunk 7 above)
Comment (fuzzybrows) [reply imports/plugins/included/product-detail-simple/client/components/productDetail.js]: I've removed this line as it was unnecessary.

File: imports/plugins/included/product-detail-simple/client/components/productDetail.js (hunk 8 above)
Comment (fuzzybrows) [reply imports/plugins/included/product-detail-simple/client/components/productDetail.js]: I've renamed the state property to cartQuantityType

File: lib/collections/collectionFS.js (hunk 9 above)
Comment (fuzzybrows) [reply lib/collections/collectionFS.js]: Done, abstracted using a function.

File: server/publications/collections/media.js (hunk 10 above)
Comment (fuzzybrows) [reply server/publications/collections/media.js]: Done, abstracted using a function

**Review by adebayo-maborukoje — APPROVED** [review]

**Pull request merged**